h2o-py/tests/testdir_algos/automl/pyunit_automl_args.py

```diff
@@ -155,48 +174,116 @@ def prostate_automl_args():
                     keep_cross_validation_fold_assignment=True)
-    aml.train(y="CAPSULE", training_frame=train)
+    aml.train(y=ds['target'], training_frame=ds['train'])
     amodel = h2o.get_model(aml.leaderboard[aml.leaderboard.nrows-1,0])
     assert amodel.params['keep_cross_validation_fold_assignment']['actual'] == False
     assert amodel._model_json["output"]["cross_validation_fold_assignment_frame_id"] == None
 
 
+def test_automl_stops_after_max_runtime_secs():
     print("Check that automl gets interrupted after `max_runtime_secs`")
     max_runtime_secs = 30
     cancel_tolerance_secs = 5   # should work for most cases given current mechanism
+    ds = import_dataset()
     aml = H2OAutoML(project_name="py_aml_max_runtime_secs", seed=1, max_runtime_secs=max_runtime_secs)
     start = time.time()
-    aml.train(y="CAPSULE", training_frame=train)
+    aml.train(y=ds['target'], training_frame=ds['train'])
     end = time.time()
     assert abs(end-start - max_runtime_secs) < cancel_tolerance_secs, end-start
 
+
+def test_stacked_ensembles_are_trained_after_timeout():
+    print("Check that Stacked Ensembles are still trained after timeout")
+    max_runtime_secs = 10
+    ds = import_dataset()
+    aml = H2OAutoML(project_name="py_aml_SE_after_timeout", seed=1, max_runtime_secs=max_runtime_secs, exclude_algos=['DeepLearning'])
+    start = time.time()
+    aml.train(y=ds['target'], training_frame=ds['train'])
+    end = time.time()
+    assert end-start - max_runtime_secs > 0
+
+    stacked_ensembles = [m for m in [aml.leaderboard[i, 0] for i in range(0, (aml.leaderboard.nrows))] if m.startswith('StackedEnsemble')]
+    assert len(stacked_ensembles) == 2, "StackedEnsemble should still be trained after timeout"
+
+
+def test_automl_stops_after_max_models():
+    print("Check that automl gets interrupted after `max_models`")
+    ds = import_dataset()
+    max_models = 5
+    aml = H2OAutoML(project_name="py_aml_max_models", seed=1, max_models=max_models)
+    aml.train(y=ds['target'], training_frame=ds['train'])
+
+    base_models = [m for m in [aml.leaderboard[i, 0] for i in range(0, (aml.leaderboard.nrows))] if not m.startswith('StackedEnsemble')]
+    assert len(base_models) == max_models, "obtained {} base models when {} are expected".format(len(base_models), max_models)
+
+
+def test_stacked_ensembles_are_trained_after_max_models():
+    print("Check that Stacked Ensembles are still trained after max models have been trained")
+    max_models = 5
+    ds = import_dataset()
+    aml = H2OAutoML(project_name="py_aml_SE_after_max_models", seed=1, max_models=max_models)
+    aml.train(y=ds['target'], training_frame=ds['train'])
+
+    stacked_ensembles = [m for m in [aml.leaderboard[i, 0] for i in range(0, (aml.leaderboard.nrows))] if m.startswith('StackedEnsemble')]
+    assert len(stacked_ensembles) == 2, "StackedEnsemble should still be trained after max models have been reached"
+
```

Comment:
Based on above it seems an `SE` will always be trained. So automl in its default form will produce `n` base models and two SE models. So in total `n + 2` models. Just adding here as a note.

Reply:
@ledell max_models count has never included SEs, cf. automl.rst above.
I created a ticket weeks ago around this (https://0xdata.atlassian.net/browse/PUBDEV-5807), ticket that I just modified to include the max_runtime_secs constraint as well.

The previous fix around max_runtimes_secs (https://0xdata.atlassian.net/browse/PUBDEV-5613) included my change where SE was included in the max_runtime_secs, and now I realize that this change was a bad idea.
I think we should be consistent there, either SE respects both early-stopping constraints (max_models + max_runtimes_secs), or none, otherwise it's confusing for everyone.
I believe the work distribution I started to implement in the XGBoost PR (there mainly to distribute models and time between various grid searches) would make it easier later if we want to consider SE just as another model family (just with higher priority).

Reply:
@seb-h2o Ok, thanks for the clarification.  When we try to fix `max_runtime_secs` to include the SEs, we can also fix `max_models` to include them at the same time.